cpp/src/arrow/dataset/filter.h

```diff
@@ -18,23 +18,413 @@
 #pragma once
 
 #include <memory>
+#include <string>
+#include <utility>
 
+#include "arrow/compute/kernels/compare.h"
+#include "arrow/dataset/type_fwd.h"
 #include "arrow/dataset/visibility.h"
+#include "arrow/result.h"
+#include "arrow/scalar.h"
 
 namespace arrow {
 namespace dataset {
 
-class ARROW_DS_EXPORT Filter {
- public:
+struct FilterType {
   enum type {
     /// Simple boolean predicate consisting of comparisons and boolean
-    /// logic (AND, OR, NOT) involving Schema fields
+    /// logic (ALL, ANY, NOT) involving Schema fields
     EXPRESSION,
 
-    ///
+    /// Non decomposable filter; must be evaluated against every record batch
     GENERIC
   };
 };
 
+class ARROW_DS_EXPORT Filter {
+ public:
+  explicit Filter(FilterType::type type) : type_(type) {}
+
+  virtual ~Filter() = default;
+
+  FilterType::type type() const { return type_; }
+
+ private:
+  FilterType::type type_;
+};
+
+/// Filter subclass encapsulating a simple boolean predicate consisting of comparisons
+/// and boolean logic (ALL, ANY, NOT) involving Schema fields
+class ARROW_DS_EXPORT ExpressionFilter : public Filter {
+ public:
+  explicit ExpressionFilter(const std::shared_ptr<Expression>& expression)
+      : Filter(FilterType::EXPRESSION), expression_(std::move(expression)) {}
+
+  const std::shared_ptr<Expression>& expression() const { return expression_; }
+
+ private:
+  std::shared_ptr<Expression> expression_;
+};
+
+struct ExpressionType {
+  enum type {
+    /// a reference to a column within a record batch, will evaluate to an array
+    FIELD,
+
+    /// a literal singular value encapuslated in a Scalar
+    SCALAR,
+
+    /// a literal Array
+    // TODO(bkietz) ARRAY,
+
+    /// an inversion of another expression
+    NOT,
+
+    /// cast an expression to a given DataType
+    // TODO(bkietz) CAST,
+
+    /// a conjunction of multiple expressions (true if all operands are true)
+    ALL,
```

Comment:
Perhaps it's only my experience, but `all` and `any` general refer to column-wide agregation, not combining individual clauses. I would expect `and` and `or` here.

Reply:
Apparently, ALL in SQL seems to correspond to my intuition: http://www.sqltutorial.org/sql-all/

@wesm  thoughts?